torchx/lib/torchx.ex

```diff
@@ -336,6 +353,12 @@ defmodule Torchx do
       {dev, ref}, other_dev when is_tensor(dev, ref) ->
         raise ArgumentError, "cannot perform operation across devices #{dev} and #{other_dev}"
 
+      [{dev, ref} | _] = tensors, nil when is_tensor(dev, ref) ->
+        prepare_tensors_list!(tensors, dev)
+
+      tensors, dev when is_list(tensors) ->
+        prepare_tensors_list!(tensors, dev)
+
```

Comment:
One alternative is to make `prepare_tensors!` recursive:

```elixir
  defp prepare_tensors!(tensors, dev \\ nil) do
    Enum.map_reduce(tensors, dev, fn
      {dev, ref}, nil when is_tensor(dev, ref) ->
        {ref, dev}

      {dev, ref}, dev when is_tensor(dev, ref) ->
        {ref, dev}

      {dev, ref}, other_dev when is_tensor(dev, ref) ->
        raise ArgumentError, "cannot perform operation across devices #{dev} and #{other_dev}"

      tensors, dev when is_list(tensors) ->
        prepare_tensors!(tensors, dev)

      bad_tensor, _dev ->
        raise ArgumentError, "expected a Torchx tensor, got: #{inspect(bad_tensor)}"
    end)
  end
```

This way you don't need to introduce `prepare_tensors_list!`. It is your call on which approach you prefer.

Reply:
The issue is that `prepare_tensors` returns a list of `{ref, dev}`, and `prepare_tensors_list` returns `{list(ref), dev}`

I'll try to make things work after all tests pass :)

Reply:
I see. So ignore me. We also don’t want to support lists of lists, so the current approach is better.